### PR TITLE
Bug 1117633 - setNFCFocusApp +autoland

### PR DESCRIPTION
**Pull request by alivedise**

### CODE REVIEW

**Comment by try-server-hook** [comment]
[alivedise Alive.Kuo (alivedise)](https://github.com/alivedise) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=6ad1552a20e8)

**Comment by try-server-hook** [comment]
[alivedise Alive.Kuo (alivedise)](https://github.com/alivedise) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=53eea6a89a43)

**Comment by try-server-hook** [comment]
[alivedise Alive.Kuo (alivedise)](https://github.com/alivedise) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=b51c6ae76637)

File: apps/system/js/browser_mixin.js [hunk 1]
```diff
@@ -57,6 +57,25 @@
     },
 
     /**
+     * This function will be called by window managers while top most app
+     * window is changed to notify nfc module in gecko.
+     */
+    setNFCFocus: function(enable) {
+      if (!this.browser || !this.browser.element ||
+          this._nfcActive === enable ||
```
Comment (etiennesegonzac) [inline apps/system/js/browser_mixin.js]: I kinda wish this case was handled by the platform (ie. calling setNFCFocus(true) twice being a no-op the second time).

File: apps/system/js/app_window_manager.js [hunk 2]
```diff
@@ -488,6 +490,13 @@
       var activeApp = this._activeApp;
       var detail = evt.detail;
       switch (evt.type) {
+        case 'hierarchytopmostwindowchanged':
+          if (Service.query('getTopMostUI').name !== this.name) {
+            return;
+          }
+          this._activeApp &&
+          this._activeApp.getTopMostWindow().setNFCFocus(true);
```
Comment (etiennesegonzac) [inline apps/system/js/app_window_manager.js]: nit: not sure about the (non)indentation

File: apps/system/js/stack_manager.js [hunk 3]
```diff
@@ -103,6 +103,7 @@ var StackManager = {
     // queueing.
     if (this._didntMove) {
       window.dispatchEvent(new CustomEvent('sheets-gesture-end'));
+      this._appOut && this._appOut._setActive(true);
```
Comment (etiennesegonzac) [inline apps/system/js/stack_manager.js]: if we're going with that we should expose a clean `setActive` method (without the `_`)

File: apps/system/js/app_window_manager.js [hunk 4]
```diff
@@ -488,6 +490,13 @@
       var activeApp = this._activeApp;
       var detail = evt.detail;
       switch (evt.type) {
+        case 'hierarchytopmostwindowchanged':
```
Comment (etiennesegonzac) [inline apps/system/js/app_window_manager.js]: do you think we'll have multiple consumer of this event or should we just listen for `windowopened`/`windowclosed` here?

File: apps/system/js/browser_mixin.js (hunk 1 above)
Comment (alivedise) [inline apps/system/js/browser_mixin.js]: agree.

File: apps/system/js/app_window_manager.js (hunk 4 above)
Comment (alivedise) [inline apps/system/js/app_window_manager.js]: For hierarchytopmostwindowchanged I think so. but I am fined to change to windowopened/windowclosed

**Comment by try-server-hook** [comment]
[alivedise Alive.Kuo (alivedise)](https://github.com/alivedise) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=330cd0c61dbc)

**Comment by try-server-hook** [comment]
[alivedise Alive.Kuo (alivedise)](https://github.com/alivedise) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=f47e4cd1579c)

**Comment by try-server-hook** [comment]
[alivedise Alive.Kuo (alivedise)](https://github.com/alivedise) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=e0fd3b33d025)

File: apps/system/test/unit/stack_manager_test.js [hunk 5]
```diff
@@ -240,6 +240,10 @@ suite('system/StackManager >', function() {
     });
   });
 
+  test('_didntMove by default is true', function() {
+    assert.isTrue(StackManager._didntMove);
```
Comment (etiennesegonzac) [inline apps/system/test/unit/stack_manager_test.js]: this is testing internal stuff, we should do our assert on something external: either the clearTransitionClasses being called or the sheets-gesture-end event not being dispatched

Comment (etiennesegonzac) [inline apps/system/test/unit/stack_manager_test.js]: or even better, asserting on the `setActive` call

File: apps/system/test/unit/app_window_test.js [hunk 6]
```diff
@@ -966,6 +977,21 @@ suite('system/AppWindow', function() {
...
 
+    test('MozBrowser API: setActive', function() {
+      var app1 = new AppWindow(fakeAppConfig1);
+      injectFakeMozBrowserAPI(app1.browser.element);
+      this.sinon.stub(app1.browser.element, 'setActive');
+      this.sinon.stub(app1.browser.element, 'setNFCFocus');
+      MockService.mTopMostUI = { name: 'Rocketbar' };
+      app1._setActive(false);
+      assert.isTrue(app1.browser.element.setActive.calledWith(false));
+      assert.isFalse(app1.browser.element.setNFCFocus.called);
```
Comment (etiennesegonzac) [inline apps/system/test/unit/app_window_test.js]: is there something preventing from assert `calledWith(false)` here?

File: apps/system/test/unit/stack_manager_test.js (hunk 5 above)
Comment (alivedise) [inline apps/system/test/unit/stack_manager_test.js]: done

**Comment by try-server-hook** [comment]
[alivedise Alive.Kuo (alivedise)](https://github.com/alivedise) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=0e17fa0dc8f6)

File: apps/system/test/unit/app_window_test.js (hunk 6 above)
Comment (alivedise) [inline apps/system/test/unit/app_window_test.js]: no, fixed

**Comment by try-server-hook** [comment]
[alivedise Alive.Kuo (alivedise)](https://github.com/alivedise) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=b5097ee87f75)

**Comment by try-server-hook** [comment]
[alivedise Alive.Kuo (alivedise)](https://github.com/alivedise) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=987ccf759c5d)

**Pull request merged**